✨ RAINCATCH-635: enhancements to load test scripts

grdryn opened the pull request:
JIRA: https://issues.jboss.org/browse/RAINCATCH-635

grdryn committed:
🔨 Pull out initialSync, do it in each flow

JIRA: https://issues.jboss.org/browse/RAINCATCH-635

I think it makes sense to have the 'sync' request in each of the
flows, since the 'syncRecords' is in there too (both should be in the
same place, wherever that is).

grdryn committed:
🚑 Fix setup script to create valid workorders

JIRA: https://issues.jboss.org/browse/RAINCATCH-635

Up to now, it was using a hardcoded workflow ID, but in the latest
version of the demo solution, those IDs are different each time.

grdryn committed:
🔥 Remove unneeded requestBodyUtils in syncDataset

JIRA: https://issues.jboss.org/browse/RAINCATCH-635

grdryn committed:
🚧 Remove confusing query_params from sync.js

JIRA: https://issues.jboss.org/browse/RAINCATCH-635

This change also adds additional meta_data properties that are at
least set for portal requests. and some TODOs too.

grdryn committed:
🚧 Pass results of sync & syncRecords in portal

JIRA: https://issues.jboss.org/browse/RAINCATCH-635

grdryn committed:
🚧 Get correct workorder in mobile flow

JIRA: https://issues.jboss.org/browse/RAINCATCH-635

grdryn committed:
🔥 Remove unused action param from createRecord

JIRA: https://issues.jboss.org/browse/RAINCATCH-635

grdryn committed:
✨ Make sync.js more useful (will be needed later)

JIRA: https://issues.jboss.org/browse/RAINCATCH-635

grdryn committed:
🔨 Pull apart sync.js for easier re-use

JIRA: https://issues.jboss.org/browse/RAINCATCH-635

We'll need to sync more than just the initial time, so it needs to be
more versatile.

grdryn committed:
🔨 Update createRecord to use sync/makeSyncBody

JIRA: https://issues.jboss.org/browse/RAINCATCH-635

grdryn committed:
🔥 Remove urlFor module

JIRA: https://issues.jboss.org/browse/RAINCATCH-635

grdryn committed:
✨ Add realistic initial sync/syncRecords dance

JIRA: https://issues.jboss.org/browse/RAINCATCH-635

Currently a client will do a sync with the old hash, then if there's
changes on the server (different hash), it will do a syncRecords with
the clientRecs that it has (none initially for us). The server will
reply with a bunch of changes that the client needs to make at this
point. The client will then do another sync (with the same old hash
that it had) and then a syncRecords with a clientRecs object that
reflects the changes that the server told it that it should apply.

This change updates the portal and mobile flows to reflect that
to-and-fro.

grdryn committed:
🔥 Remove unused syncResultsToArray module

JIRA: https://issues.jboss.org/browse/RAINCATCH-635

grdryn committed:
🎨 Stop shadowing the return value of createRecord

JIRA: https://issues.jboss.org/browse/RAINCATCH-635

grdryn committed:
🎨 Improve formatting before changing

JIRA: https://issues.jboss.org/browse/RAINCATCH-635

grdryn committed:
🚧 WIP: cleanup

JIRA: https://issues.jboss.org/browse/RAINCATCH-635

grdryn committed:
✨ Add sync loop with interval to createRecord

JIRA: https://issues.jboss.org/browse/RAINCATCH-635

This is just to make sure that we wait until we hear from the server
that it has applied our update

grdryn committed:
🚧 Add a function to update clientRecs

JIRA: https://issues.jboss.org/browse/RAINCATCH-635

grdryn committed:
🚧 Add a bund of TODOs to mobile flow script

JIRA: https://issues.jboss.org/browse/RAINCATCH-635

grdryn commented:
@david-martin This is how far I've gotten with this. You'll see a bunch of TODOs where things need to be done. Also, when I started, I was updating the mobile and portal flows in tandem, but the most recent changes are only on the mobile flow. Once they're finished they'll also need to be reflected in the portal flow.

I might be on for a while later, so feel free to leave me any questions here and I'll try to get back to them. I won't be online tomorrow! :)

cc @paolobueno

david-martin commented:
@grdryn  @paolobueno I'm having trouble running this locally.
I'll try dig in at a later time.
Here's the errors from the demo cloud app i see if you have any pointers
https://gist.github.com/david-martin/b31cc481d7f62289a012b20362b91531

grdryn commented on scripts/mobile-flow.js:
```diff
@@ -32,30 +32,31 @@ module.exports = function mobileFlow(runner, argv, clientId) {
       // Then do another syncRecords, this time with clientRecs
-        .spread((syncResults, syncRecordsResults) => Promise.all([
+        .spread((syncResults, initialSyncRecordsResponses) => Promise.all([
+          // TODO: in the browser, I see a hash in the response from syncRecords, but not here in the script
```
@david-martin this TODO is something that could be investigated separately to the improvements to the scripts that's being done here.

grdryn commented on util/fixtures/makeWorkorder.js:
```diff
@@ -15,6 +15,7 @@ module.exports = function makeWorkorder(userId, workflowId) {
     finishDate: "2029-02-23T00:00:00Z",
     finishTime: "2029-02-23T00:00:00Z",
     summary: "Do the work",
+    // TODO: for some reason 'startTimestamp' now shows up as 'finish date/time' in the portal
```
@david-martin also this is a separate issue that could be looked at in the raincatcher demo itself (not sure if it's considered a bug or a feature)

cc @paolobueno

grdryn replied:
@paolobueno also, given that it's `startTimestamp` that seems to populate those fields now, should it be set to some time in the future, like `finishDate` and `finishTime` were above?

paolobueno replied:
I guess there's no harm in setting it forward. Easy to revert when the weirdness around the sematincs of these fields is resolved and eventually validation is implemented.

grdryn committed:
🚧 WIP: Handle clientRecs better around syncRecords calls

JIRA: https://issues.jboss.org/browse/RAINCATCH-635

grdryn committed:
🐛 Fix setup script after other changes

JIRA: https://issues.jboss.org/browse/RAINCATCH-635

grdryn committed:
✨ Allow passing of action through createRecord

JIRA: https://issues.jboss.org/browse/RAINCATCH-635

grdryn committed:
🚧 Make a working mobile flow

JIRA: https://issues.jboss.org/browse/RAINCATCH-635

The changes here (of which there are many), give a working mobile
flow. By 'working', I mean that the steps don't collide with each
other, and the one result record is updated continuously. After
running this flow, the workorder should be marked as completed if you
look at it in the portal.

grdryn committed:
🚧 Make portal flow work again with new changes

JIRA: https://issues.jboss.org/browse/RAINCATCH-635

grdryn committed:
🐛 Set startTimestamp to future since it's finish

JIRA: https://issues.jboss.org/browse/RAINCATCH-635

grdryn committed:
✨ Add query_params to mobile flow as in demo app

JIRA: https://issues.jboss.org/browse/RAINCATCH-635

grdryn committed:
✨ Sync with acknowledgements in mobile flow

JIRA: https://issues.jboss.org/browse/RAINCATCH-635

grdryn committed:
✨ Sync with acknowledgements in portal flow

JIRA: https://issues.jboss.org/browse/RAINCATCH-635

grdryn committed:
🔨 Pull out syncRecords>ack>syncRecords>ack flow

JIRA: https://issues.jboss.org/browse/RAINCATCH-635

grdryn committed:
💄 Make it easier to identify generated users

JIRA: https://issues.jboss.org/browse/RAINCATCH-635

paolobueno commented on scripts/mobile-flow.js:
```diff
@@ -118,10 +113,26 @@ module.exports = function mobileFlow(runner, argv, clientId) {
+                          Promise.resolve(result),
+                          Promise.resolve(_.find(
+                            _.get(doSyncRecordsResult, 'res.create', {}),
+                            r => r.data.id === _.map(result.updates.applied, a => a.uid)[0]
```
can use `_.find` here to return the first element instead of creating a full new array.

paolobueno commented on scripts/mobile-flow.js:
```diff
@@ -118,10 +113,26 @@ module.exports = function mobileFlow(runner, argv, clientId) {
+                  .spread((syncResponse, updatedRecord, clientRecs) => Promise.all([
+                    doSync(`${baseUrl}/mbaas/sync/result`, makeSyncBody('result', clientId, syncResponse.hash, queryParams(user.id).result, null, _.values(_.get(syncResponse, 'updates.applied')))),
+                    Promise.resolve(updatedRecord),
+                    Promise.resolve(clientRecs)
```
is this `.all` call needed? the return value doesn't seem to be used.

paolobueno commented on scripts/mobile-flow.js:
```diff
@@ -69,8 +69,7 @@ module.exports = function mobileFlow(runner, argv, clientId) {
           .spread((syncResponse, createdRecord, clientRecs) => Promise.all([
-            // TODO: sync with acknowledgements (still old dataset hash)
-            Promise.resolve(syncResponse),
+            doSync(`${baseUrl}/mbaas/sync/result`, makeSyncBody('result', clientId, syncResponse.hash, queryParams(user.id).result, null, _.values(_.get(syncResponse, 'updates.applied')))),
```
the `doSync` calls here have become pretty complex, would be nice to have a little function to clean up and DRY the invocation.
